### PR TITLE
Fix #8816 (Dynlink/packing issue in bytecode with 4.08)

### LINKED ISSUE #8816: Dynlink/packing issue in bytecode with 4.08
With 4.08 (including the 4.08.1+rc1), we have an issue with the bytecode version of Frama-C, where the dynamic loading of modules (via Dynlink) no longer works as before, with error messages such as "no implementation available for <module>".

A quick glance through the changelog did not indicate a likely culprit. We had a [typing issue related to -pack](https://github.com/ocaml/ocaml/issues/7932) which has already been solved in our development version, but this bug might be unrelated, since we can now type, compile and run Frama-C in native version without issues.

For a more concrete description, here's an example of what happens when trying to run `frama-c.byte` after compiling it, in 4.08 (among several similar messages):

```
[kernel] User Error: cannot load plug-in 'Callgraph': cannot load module
  Details: no implementation available for Callgraph.Options
```

`Callgraph.cmo` is produced by packing several modules, including an `options.cmo`. This compiles, both in bytecode and in native version, and when running the native version of Frama-C (which loads the corresponding `.cmxs` file), everything works fine. However, when running the bytecode version (`frama-c.byte`), these errors appear. Both versions work fine up to 4.07. (I could not test with 4.09 since we depend on `dune`, but I assume no changes related to it.)

We can try to produce a minimal, self-contained version, but it could take some time. Do you have an idea of something else we could do to further debug this issue?

### PR DESCRIPTION
**Pull request by lpw25**
This fixes #8816 by only including toplevel modules in the list of implementation imports of `.cmo` files.

### CODE REVIEW

File: otherlibs/dynlink/byte/dynlink.ml [hunk 1]
```diff
@@ -39,7 +39,11 @@ module Bytecode = struct
         @ Symtable.required_globals t.cu_reloc
       in
       let required =
-        List.filter (fun id -> not (Ident.is_predef id)) required
+        List.filter
+          (fun id ->
+             not (Ident.is_predef id)
+             && not (String.contains (Ident.name id) '.'))
```
Comment (gasche) [inline otherlibs/dynlink/byte/dynlink.ml:45]: Without knowing the context, this looks a bit strange. Where is the code that creates predefined identifiers with `.` in their name?

Comment (gasche) [reply otherlibs/dynlink/byte/dynlink.ml:45]: (Well I guess those are the `Ident.create_persistent (packagename ^ "." ^ name)` calls in `bytecomp/bytepackager.ml`.)

Comment (lpw25) [reply otherlibs/dynlink/byte/dynlink.ml:45]: Yeah, and you'll also notice a use of `String.contains ... '.'` in that file as well. It's not pleasant, but there are a couple of PRs coming down the pipeline that will tidy up this stuff so I'm not in a hurry to fix it right now.

**Review by gasche — APPROVED** [review]
The fix is not beautiful but it is simple and well-tested -- and it works. I agree with @lpw25's approach of getting the simplest fix out first, and then making things more robust in trunk.

**Comment by dra27** [comment]
I was going to push a `Changes` entry, but it's actually non-obvious where it goes!

@Octachron - given the report against 4.08.1+rc1, do you want this on 4.08?

@gasche - there's no section in `trunk` Changes for 4.08.1 - or should the numbers just be added to the original Dynlink soundness entry in 4.08.0, and put in different sections on the 4.08 and 4.09 branches??

**Comment by gasche** [comment]
Yes, we want this in 4.08.1. We have a "4.08 maintenance branch" section in the 4.08 branch, we need to import it in the other banches. Let me take care of that right now.

**Comment by gasche** [comment]
(Wouldn't it be nice if we had tooling support to more easily work with Changes entries? Just kidding.)

**Comment by dra27** [comment]
You kid - I did a few months ago start to resurrect the tool I wrote in 2017, but it needs more than an hour or so's work to resurrect, so it sits on the spike for now until I have something *really* important I should be doing and need distracting from :wink:

**Comment by gasche** [comment]
I put the Changes entries in order, all post-4.08 branches should now have a "4.08 maintenance" section.

**Comment by dra27** [comment]
Thanks @gasche - I've rebased and pushed a Changes entry, so I think is ready squash and pick once CI reconfirms.

**Comment by gasche** [comment]
If you allow a very minor nitpick, "Fix #N" is not a very good name for a PR. When I have to find them again in the PR list, such titles make me unhappy. I edited this one and the other one (I just repeated/summarized the name of the bug they are fixing); let's avoid them in the future.

**Comment by XVilka** [comment]
By the way, it makes sense to suggest GitHub autocomplete the PR title with the issue title itself, once you type 'Fix #xxxx' in the PR title. Would solve exactly this problem.

**Pull request merged**